<!-- 
    Launches:
        Map Server
        AMCL
        TF Listener
 -->

 <launch>
    <param name="use_sim_time" value="true"/>

    <arg name="robot_name" default="gopher"/>
    <arg name="prefix" default="$(arg robot_name)/"/>

    <arg name="cmd_vel" default="$(arg prefix)cmd_vel"/>
    <arg name="scan" default="$(arg prefix)base_scan"/>  <!--Used-->
    <arg name="odom_frame" default="$(arg prefix)odom"/> <!--Used-->
    <arg name="base_frame" default="$(arg prefix)chassis_link"/> <!--Used-->
    <arg name="laser_frame" default="$(arg prefix)laser_link"/>

    <!-- map for the map server -->
    <arg name="map_name" default="hospital_floorplan_objects"/> <!--Used-->

    <!-- Map loader -->
    <node name="map_server" pkg="map_server" type="map_server" 
          args="$(find gopher_navigation)/maps/$(arg map_name).yaml">
    </node>


    <!-- AMCL Node -->
    <node name="amcl" pkg="amcl" type="amcl" output="screen">
        <!-- Filter -->
        <!-- <param name="min_particles" value="20"/>
        <param name="max_particles" value="200"/>
        <param name="update_min_a" value="0.1"/>
        <param name="update_min_d" value="0.25"/> -->
        <param name="min_particles"             value="200"/>
        <param name="max_particles"             value="500"/>
        <param name="kld_err"                   value="0.02"/>
        <param name="update_min_d"              value="0.20"/>
        <param name="update_min_a"              value="0.20"/>
        <param name="resample_interval"         value="1"/>
        <param name="transform_tolerance"       value="0.5"/>
        <param name="recovery_alpha_slow"       value="0.00"/>
        <param name="recovery_alpha_fast"       value="0.00"/>
        <param name="gui_publish_rate"          value="50.0"/>

        <!-- Laser -->
        <remap from="scan" to="$(arg scan)"/>
        <param name="laser_max_range"           value="5.0"/>
        <param name="laser_max_beams"           value="180"/>
        <param name="laser_z_hit"               value="0.5"/>
        <param name="laser_z_short"             value="0.05"/>
        <param name="laser_z_max"               value="0.05"/>
        <param name="laser_z_rand"              value="0.5"/>
        <param name="laser_sigma_hit"           value="0.2"/>
        <param name="laser_lambda_short"        value="0.1"/>
        <param name="laser_likelihood_max_dist" value="2.0"/>
        <param name="laser_model_type"          value="likelihood_field"/>

        <!-- Odometry -->
        <param name="odom_frame_id" value="$(arg odom_frame)"/>
        <param name="odom_model_type" value="diff-corrected"/>
        <param name="base_frame_id" value="$(arg base_frame)"/>
        <param name="global_frame_id" value="map"/>
        <param name="odom_alpha1"               value="0.005"/>
        <param name="odom_alpha2"               value="0.005"/>
        <param name="odom_alpha3"               value="0.005"/>
        <param name="odom_alpha4"               value="0.005"/>
    </node>

    <node name="tf_listener" pkg="gopher_navigation"  type="tf_listener_node.py" respawn="true" output="screen"/>

</launch>
